Tella/Data/Networking/Repositories/GDriveRepository.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  GDriveRepository.swift
+//  Tella
+//
+//  Created by gus valbuena on 5/28/24.
+//  Copyright © 2024 HORIZONTAL. All rights reserved.
+//
+
+import Foundation
+import GoogleSignIn
+import GoogleAPIClientForREST
+
+protocol GDriveRepositoryProtocol {
+    func handleSignInButton(completion: @escaping (Result<Void, Error>) -> Void)
+    func restorePreviousSignIn(completion: ((_ user: GIDGoogleUser?) -> Void)?)
+    func handleUrl(url: URL)
+    func getSharedDrives(googleUser: GIDGoogleUser?, completion: @escaping (Result<[GTLRDrive_Drive], Error>) -> Void)
+}
+
+struct GDriveRepository: GDriveRepositoryProtocol  {
+    static let shared = GDriveRepository()
+    private var rootViewController: UIViewController? {
+        return UIApplication.shared.windows.first?.rootViewController
+    }
+    
+    func handleSignInButton(completion: @escaping (Result<Void, Error>) -> Void) {
+        guard let rootViewController = self.rootViewController else {
+            print("There is no root view controller!")
+            return
+        }
+        
+        GIDSignIn.sharedInstance.signIn(withPresenting: rootViewController) { signInResult, error in
+            if let error = error {
+                completion(.failure(error))
+                return
+            } else {
+                completion(.success(()))
+            }
+        }
+    }
+    
+    func restorePreviousSignIn(completion: ((_ user: GIDGoogleUser?) -> Void)? = nil) {
```

Comment:
What do you think if we return a domain model that contains only the parameters that we need instead of `GIDGoogleUser` ?

Reply:
yeah makes sense. I'm wondering if it would be better to not return the GIDGoogleUser and have it only as a state inside the repository. Because we don't need it outside of the repository actually. We use it to pass it as the authorizer for google drive calls. What do you think of this?

For the `GTLRDrive_Drive` I think it makes more sense!

Reply:
yeah you're right, I think you've resolved that by adding the `googleUser` as a private ?

Reply:
yeah! so now we handle the token inside the repository so the rest of the app doesn't have that responsability